crates/compiler/load_internal/src/file.rs

```diff
@@ -5505,31 +5163,255 @@ fn canonicalize_and_constrain<'a>(
+                // created an IdentId for this, when it was imported exposed
+                // in a dependent module.
+                //
+                // For example, if module A implements [B.{ foo }], then
+                // when we get here for B, `foo` will already have
+                // an IdentId. We must reuse that!
+                let ident_id = ident_ids.get_or_insert(loc_name.value.as_str());
+                let symbol = Symbol::new(header.module_id, ident_id);
+
+                exposed.push(symbol);
+            }
+        }
+
+        if cfg!(debug_assertions) {
+            header.module_id.register_debug_idents(ident_ids);
+        }
+
+        ident_ids.clone()
+    };
+
+    // This module depends on all the modules it exposes. We need to load those in order
+    // to do things like report errors for them, generate docs for them, etc.
+    if header.is_root_module {
+        if let HeaderType::Platform {
+            exposes,
+            exposes_ids,
+            ..
+        }
+        | HeaderType::Package {
+            exposes,
+            exposes_ids,
+            ..
+        } = header.header_type
+        {
+            for (loc_module_name, module_id) in exposes.iter().zip(exposes_ids.iter().copied()) {
+                let module_name_str = loc_module_name.value.as_str();
+                let pq_module_name = PackageQualified::Unqualified(module_name_str.into());
+
+                // We should never change an entry here. Either we should have no entry,
+                // or if we do have one, it should be unchanged by this insertion.
+                debug_assert_eq!(
+                    &module_id,
+                    deps_by_name.get(&pq_module_name).unwrap_or(&module_id),
+                    "Already had a deps_by_name entry for {:?}, but it was {:?} rather than the expected {:?}",
+                    pq_module_name,
+                    deps_by_name.get(&pq_module_name).unwrap(),
+                    module_id,
+                );
+                deps_by_name.insert(pq_module_name, module_id);
+            }
+        }
+    }
+
+    // Send the deps to the coordinator thread for processing,
+    // then continue on to parsing and canonicalizing defs.
+    //
+    // We always need to send these, even if deps is empty,
+    // because the coordinator thread needs to receive this message
+    // to decrement its "pending" count.
+    let mut package_qualified_imported_modules = MutSet::default();
+    for (pq_module_name, module_id) in &deps_by_name {
+        match pq_module_name {
+            PackageQualified::Unqualified(_) => {
+                package_qualified_imported_modules
+                    .insert(PackageQualified::Unqualified(*module_id));
+            }
+            PackageQualified::Qualified(shorthand, _) => {
+                package_qualified_imported_modules
+                    .insert(PackageQualified::Qualified(shorthand, *module_id));
+            }
+        }
```

Comment:
these branches are equivalent except for the packing/unpacking. consider adding a `PackageQualified#map_module` method.